pkg/chunk/aws/s3_storage_client.go

```diff
@@ -53,43 +74,133 @@ func (cfg *S3Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 	}
-	s3Config, err := awscommon.ConfigFromURL(cfg.S3.URL)
+
+	sess, err := session.NewSession(s3Config)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("Failed to create new s3 session: %v", err)
 	}
 
-	s3Config = s3Config.WithS3ForcePathStyle(cfg.S3ForcePathStyle) // support for Path Style S3 url if has the flag
+	s3Client := s3.New(sess)
+
+	var sseEncryption *string
+	if cfg.SSEEncryption {
+		sseEncryption = aws.String("AES256")
+	}
+
+	client := S3ObjectClient{
+		S3:              s3Client,
+		bucketNames:     bucketNames,
+		delimiter:       delimiter,
+		sseEncryption:   sseEncryption,
+		putUserMetadata: cfg.PutUserMetadata,
+	}
+	return &client, nil
+}
+
+func buildS3Config(cfg S3Config) (*aws.Config, []string, error) {
+	var s3Config *aws.Config
+	var err error
+
+	// if an s3 url is passed use it to initialize the s3Config and then override with any additional params
+	if cfg.S3.URL != nil {
+		s3Config, err = awscommon.ConfigFromURL(cfg.S3.URL)
+		if err != nil {
+			return nil, nil, err
+		}
+	} else {
+		s3Config = &aws.Config{}
+		s3Config = s3Config.WithRegion("dummy")
+		s3Config = s3Config.WithCredentials(credentials.AnonymousCredentials)
+	}
 
 	s3Config = s3Config.WithMaxRetries(0) // We do our own retries, so we can monitor them
 	s3Config = s3Config.WithHTTPClient(&http.Client{Transport: defaultTransport})
-	sess, err := session.NewSession(s3Config)
-	if err != nil {
-		return nil, err
+	s3Config = s3Config.WithS3ForcePathStyle(cfg.S3ForcePathStyle) // support for Path Style S3 url if has the flag
+
+	if cfg.Endpoint != "" {
+		s3Config = s3Config.WithEndpoint(cfg.Endpoint)
 	}
-	s3Client := s3.New(sess)
-	bucketNames := []string{strings.TrimPrefix(cfg.S3.URL.Path, "/")}
+
+	if cfg.Insecure {
+		s3Config = s3Config.WithDisableSSL(true)
+	}
+
+	if cfg.Region != "" {
+		s3Config = s3Config.WithRegion(cfg.Region)
+	}
+
+	if cfg.AccessKeyID != "" && cfg.SecretAccessKey == "" ||
+		cfg.AccessKeyID == "" && cfg.SecretAccessKey != "" {
+		return nil, nil, errors.New("Must supply both an Access Key ID and Secret Access Key or neither")
+	}
+
+	if cfg.AccessKeyID != "" && cfg.SecretAccessKey != "" {
+		creds := credentials.NewStaticCredentials(cfg.AccessKeyID, cfg.SecretAccessKey, "")
+		s3Config = s3Config.WithCredentials(creds)
+	}
+
+	// Use a custom http.Client with the golang defaults but also specifying
+	// MaxIdleConnsPerHost because of a bug in golang https://github.com/golang/go/issues/13801
+	// where MaxIdleConnsPerHost does not work as expected.
+	s3Config = s3Config.WithHTTPClient(&http.Client{
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+				DualStack: true,
+			}).DialContext,
+			MaxIdleConns:          100,
+			IdleConnTimeout:       cfg.HTTPConfig.IdleConnTimeout, //90 * time.Second,
+			MaxIdleConnsPerHost:   100,
+			TLSHandshakeTimeout:   3 * time.Second,
```

Comment:
Above you mentioned we use "golang defaults", but this is not the default.

Reply:
Adjusted the comment to clarify where that config came from and why it exists.